what-is-chalk.mdx

```diff
@@ -18,215 +18,430 @@ import { PyDiffEditor } from '@/components/Editor'
+- serving fresh features on-the-fly with versioning, branching, and full observability
+    - gradually rollout, easily rollback, and feature flag models with version control
+    - collaborate across teams with branch-based QA
+    - A/B test with real historical production traffic by leveraging Chalk’s feature store
 
-Chalk makes it simple to develop feature pipelines for machine
-learning. Define Python functions using the libraries and tools you're
-familiar with instead of specialized DSLs. Chalk then orchestrates
-your functions into pipelines that execute in parallel on a Rust-based
-engine and coordinates the infrastructure required to compute
-features.
+## How does Chalk make feature engineering easier?
 
-### Features
+Chalk lets you define features using Pythonic dataclasses–just like Pydantic–no DSLs.
+Every feature is:
+- typed and validated
+- versioned and testable
+- composable and reusable in both training and online inference
 
-To get started, [define your features](/docs/features) with
-[Pydantic](https://pydantic-docs.helpmanual.io/)-inspired Python classes.
-You can define schemas, specify relationships, and add metadata
-to help your team share and re-use work.
+Describe relationships between entities (e.g. users and transactions) with simple type annotations and Chalk takes care of joins, lineage, and query planning automatically.
 
 ```py
-from chalk import _, DataFrame
 import chalk.functions as F
-from chalk.features import features
+from chalk import features, DataFrame, Windowed, windowed, _
 
 @features
 class Transaction:
     id: int
     amount: float
+    discount_percentage: float
+
+    # create new features inline with Chalk Expressions
+    is_expensive_purchase: bool = _.amount > 100
+
+    # instead of user_id being a str type -> user_id: str
+    # we can point to a feature from another class
     user_id: "User.id"
+
+    # Chalk infers the join key enabling us to reference
+    # the User associated with this transaction!
     user: "User"
 
 @features
 class User:
+    # id, name, email are pulled from underlying data sources like
+    # SQL (Postgres, Athena, Glue, Iceberg), Snowflake, Databricks, Kafka streams, etc.
     id: int
     name: str
     email: str
+
+    # computed via a Python resolver, more on this later
+    username: str
+
+    # Chalk SDK provides common ML primitives
     name_match: float = F.levenshtein_distance(_.name, _.email)
+
+    # Leverage "User.id" type annotation from Transaction class
+    # as a join key to create intermediary dataframes
     txns: DataFrame[Transaction]
-    count_txns: Windowed[int] = windowed(
-        "1m", "1d", "30d",
-        _.txns.count()
+
+    # reference dataframes to run aggregations
+    transaction_count: int = _.txns.count()
+
+    # easily filter dataframes before running an aggregation
+    average_discount_percentage: float = _.txns[_.discount_percentage > 0].mean()
```

Comment:
this is missing a field on which the aggregation is being performed. probably needs to project out the amount or something like that

Reply:
done!